### PR TITLE
Make solver able to reject build-type: Custom packages

### PR DESCRIPTION
**Pull request by phadej**
---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by grayjay** [comment]
Could you please add more background on this flag to the commit message?

**Comment by phadej** [comment]
@grayjay will do once the idea is in general accepted (along as writing docs and the changelog).

**Comment by gbaz** [comment]
@phadej to accept the idea in general would probably require a little motivation -- what's the plan for why we'd want this and how it would be used?

**Comment by phadej** [comment]
`build-type: Custom` packages can run arbitrary code (technically `build-type: Configure` packages too). IMO it's fair to be able to ask solver to avoid them. Probably we'll need a more granular constraints, to allow *some* `build-type: Custom` packages, but this is a start.

**Comment by gbaz** [comment]
Is there a concrete use case for this? Are there consumers of cabal-install we can imagine wanting to turn this on by default in certain of their build systems or the like, or is the motivation more "this should exist" and somewhat speculative? Like the code looks clean to me, and the approach reasonable. I'm just not clear on if there's a general direction of work this is considered a part of or etc?

**Comment by phadej** [comment]
>  Are there consumers

Myself. E.g. for GHCJS (or crosscompiling if it worked).

**Comment by gbaz** [comment]
Thanks, ghcjs was the sort of motivating use case that I was asking about! Being able to rule out plans ghcjs can't use makes perfect sense.

**Comment by phadej** [comment]
GHCJS can use `build-type: Custom` packages (fortunately!) but things are smoother when these are avoided

**Review by gbaz — APPROVED** [review]

File: cabal-install-solver/src/Distribution/Solver/Modular/IndexConversion.hs [hunk 1]
```diff
@@ -344,6 +344,7 @@ convCondTree flags dr pkg os arch cinfo pn fds comp getInfo solveExes@(SolveExec
               ++ L.map (\e -> D.Simple (LDep dr (Ext  e)) comp) (allExtensions bi) -- unconditional extension dependencies
               ++ L.map (\l -> D.Simple (LDep dr (Lang l)) comp) (allLanguages  bi) -- unconditional language dependencies
               ++ L.map (\(PkgconfigDependency pkn vr) -> D.Simple (LDep dr (Pkg pkn vr)) comp) (pkgconfigDepends bi) -- unconditional pkg-config dependencies
+              ++ [D.Simple (LDep dr (BT (buildType pkg))) comp]
```
Comment (grayjay) [inline cabal-install-solver/src/Distribution/Solver/Modular/IndexConversion.hs:347]: This line seems to add the build type constraint to every component and the branches of every conditional. Could the constraint be associated with the whole package instead, similar to UnsupportedSpecVer?

File: cabal-install-solver/src/Distribution/Solver/Types/PackageConstraint.hs [hunk 2]
```diff
@@ -141,8 +141,8 @@ showPackageConstraint pc@(PackageConstraint scope prop) =
...
-    toDep (PackagePropertyVersion vr) = Just $ PackageVersionConstraint (scopeToPackageName scope) vr
-    toDep (PackagePropertyInstalled)  = Nothing
-    toDep (PackagePropertySource)     = Nothing
-    toDep (PackagePropertyFlags _)    = Nothing
-    toDep (PackagePropertyStanzas _)  = Nothing
+    toDep (PackagePropertyVersion vr)  = Just $ PackageVersionConstraint (scopeToPackageName scope) vr
+    toDep (PackagePropertyInstalled)   = Nothing
+    toDep (PackagePropertySource)      = Nothing
+    toDep (PackagePropertyFlags _)     = Nothing
+    toDep (PackagePropertyStanzas _)   = Nothing
```
Comment (grayjay) [inline cabal-install-solver/src/Distribution/Solver/Types/PackageConstraint.hs:148]: It looks like this file has only whitespace changes.

**Comment by phadej** [comment]
Cleaning up old PRs.

**Pull request closed without merging**